File: ios/Pipeline/ImagePipeline.swift
import UIKit

struct ImageResizeOptions {

    let input: String
    let output: String

    let width: Double?
    let height: Double?

    let quality: Double?
    let format: ImageFormat?
    let keepAspectRatio: Bool

    init(
        input: String,
        output: String,
        width: Double? = nil,
        height: Double? = nil,
        quality: Double? = nil,
        format: ImageFormat? = nil,
        keepAspectRatio: Bool = true
    ) {
        self.input = input
        self.output = output
        self.width = width
        self.height = height
        self.quality = quality
        self.format = format
        self.keepAspectRatio = keepAspectRatio
    }
}

enum ImagePipeline {

    static func execute(
        options: ImageResizeOptions
    ) throws -> ImageResult {

        // Decode
        var image = try Decoder.decode(
            options.input
        )

        // Fix EXIF orientation
        image = ExifUtils.fixOrientation(
            image
        )

        // Resize
        image = Resize.execute(
            image,
            options: options
        )

        // Resolve Nitro optional values
        let format = options.format ?? .jpeg
        let quality = options.quality ?? 90.0

        // Prepare storage
        let store = try ImageStore.prepare(
            format: format
        )

        // Encode cache
        let size = try Encoder.encode(
            image: image,
            output: store.cacheFile,
            format: format,
            quality: quality
        )

        // Persist permanent file
        try ImageStore.persist(
            store
        )

        return ImageResult(
            id: store.id,
            path: store.cacheFile.path,
            permanentPath: store.permanentFile.path,
            width: image.size.width,
            height: image.size.height,
            size: Double(size)
        )
    }
}

File: android/src/main/java/com/margelo/nitro/spiralimage/pipeline/ImagePipeline.kt
package com.margelo.nitro.spiralimage.pipeline

import com.margelo.nitro.spiralimage.*
import com.margelo.nitro.spiralimage.store.CacheManager
import com.margelo.nitro.spiralimage.store.ImageIdGenerator
import java.io.File

object ImagePipeline {

    fun process(
        cacheDir: File,
        options: ProcessOptions
    ): ImageResult {

        val start = System.currentTimeMillis()

        val inputPath =
            options.path.removePrefix("file://")

        val id =
            ImageIdGenerator.generate(inputPath)

        val outputConfig =
            options.output

        val resizeConfig =
            options.resize

        val extension = when (outputConfig?.format) {
            ImageFormat.PNG -> "png"
            ImageFormat.WEBP -> "webp"
            else -> "jpg"
        }

        val outputFile =
            outputConfig?.path
                ?.let { File(it.removePrefix("file://")) }
                ?: CacheManager.createFile(
                    cacheDir,
                    id,
                    extension
                )

        val bitmap =
            Decoder.decode(inputPath)

        val resized =
            Resize.execute(
                bitmap,
                resizeConfig?.width?.toInt(),
                resizeConfig?.height?.toInt()
            )

        val size =
            Encoder.encode(
                resized,
                outputFile,
                outputConfig?.format ?: ImageFormat.JPEG,
                outputConfig?.quality?.toInt() ?: 90
            )

        if (bitmap !== resized) {
            bitmap.recycle()
        }
        resized.recycle()

        val _elapsed = System.currentTimeMillis() - start

        return ImageResult(
            id = id,
            path = outputFile.absolutePath,
            permanentPath = outputFile.absolutePath,
            width = resized.width.toDouble(),
            height = resized.height.toDouble(),
            size = size
        )
    }
}